mgr/dashboard_v2: add filesystem page

tspmelo opened the pull request:
This PR will bring the filesystem and clients pages to the dashboard v2.

In this version I merged the clients page into the filesystem page, removing the need to open a new page to see the clients list.

It is still missing unit tests and is missing some pipes and services from #31. 
The menu is not working. If you wish to test it you will need to open the following url: `/#/filesystem/1`

current screenshot:
![192 168 1 136_4200_ fhd 3](https://user-images.githubusercontent.com/399326/35937457-2e269a64-0c3e-11e8-9472-7f2136f1e3f9.png)

`Signed-off-by: Tiago Melo <tmelo@suse.com>`

sebastian-philipp commented on src/pybind/mgr/dashboard_v2/controllers/filesystem.py:
```diff
@@ -0,0 +1,357 @@
+LOG_BUFFER_SIZE = 30
+
+
+@ApiController('filesystem')
```
I'd call this `cephfs` instead.

tspmelo replied:
done

sebastian-philipp commented on src/pybind/mgr/dashboard_v2/controllers/filesystem.py:
```diff
@@ -0,0 +1,357 @@
+            fs_id = int(fs_id)
+        except ValueError:
+            raise cherrypy.HTTPError(400,
+                                     "Invalid filesystem id {0}".format(fs_id))
```
```Python
def fs_id_to_int(fs_id):
    try:
        return int(fs_id)
    except ValueError:
        raise cherrypy.HTTPError(400, "Invalid cephfs id {}".format(fs_id))
```

tspmelo replied:
done

sebastian-philipp commented on src/pybind/mgr/dashboard_v2/controllers/filesystem.py:
```diff
@@ -0,0 +1,357 @@
+    @cherrypy.tools.json_out()
+    def mds_counters(self, fs_id):
+        """
+        Result format: map of daemon name to map of counter to list of datapoints
```
You can help our type checker by adding this to the docstring: `:rtype: dict[str, dict[str, list]]`

tspmelo replied:
done

sebastian-philipp commented on src/pybind/mgr/dashboard_v2/controllers/filesystem.py:
```diff
@@ -0,0 +1,357 @@
+
+        return names
+
+    def get_rate(self, daemon_type, daemon_name, stat):
```
is `daemon_type` actually `svc_type` ?

sebastian-philipp commented on src/pybind/mgr/dashboard_v2/controllers/filesystem.py:
```diff
@@ -0,0 +1,357 @@
+        else:
+            return 0
+
+    def format_dimless(self, n, width, colored=True):
```
1. The API should not care about colours. I'd just remove it. 
2. The API should not care about formatting values in general. I'd just remove the whole method. That's a job for the frontend.

tspmelo replied:
done

sebastian-philipp commented on src/pybind/mgr/dashboard_v2/controllers/filesystem.py:
```diff
@@ -0,0 +1,357 @@
+            return formatted
+
+
+class CephFSClients(object):
```
No need to change this, but please have a look at https://www.youtube.com/watch?v=o9pEzgHorH0

sebastian-philipp commented on src/pybind/mgr/dashboard_v2/controllers/filesystem.py:
```diff
@@ -0,0 +1,357 @@
+    def get(self):
+        mds_spec = "{0}:0".format(self.fscid)
+        result = CommandResult("")
+        self._module.send_command(result, "mds", mds_spec,
```
If possible I wouldn't directly use self._module.send_command, as it is awkward to use. E.g. we have `models/send_command_api.py`

sebastian-philipp commented on src/pybind/mgr/dashboard_v2/controllers/cephfs.py:
```diff
@@ -0,0 +1,321 @@
+
+        return dict(result)
+
+    def fs_id_to_int(self, fs_id):
```
As you've added this to the class, it can be marked as `@staticmethod`

tspmelo replied:
done

jcsp commented:
FYI, one reason for the separate clients page was the extra cost of fetching that information.  The overall filesystem status is local to the mgr, but getting the list of clients involves sending a command to the MDS to fetch it.

The other reason is just that the clients can be very numerous, so I think people will still want a full-sized clients page, even if a little table is also included in the filesystem status page.

Commit:
mgr/dashboard_v2: add filesystem page


Signed-off-by: Tiago Melo <tmelo@suse.com>

tspmelo commented:
This page is now fully functional.

Changes: 
- all tables are now using cdTable and cdTableKeyValue;
- converted the standby table into a key value table
- implemented the recommendations from @sebastian-philipp ; 
- implemented the recommendations from @jcsp; 

New Filesystem page:
![localhost_4200_ fh](https://user-images.githubusercontent.com/399326/36318898-b51ec2b4-1330-11e8-8df3-d1a1d2d3a5ca.png)

New Clients page:
![localhost_4200_ fh 1](https://user-images.githubusercontent.com/399326/36318899-b65a9b1c-1330-11e8-9767-2cbd1f8c25b1.png)

LenzGr reviewed (dismissed):
Nice work! Thanks.

votdev commented:
Shouldn't we adapt the oA workflow here and display the file systems in a datatable instead of showing the available file systems in the navigation menu dropdown?

In this case the Pools implementation must be adapted, too.

votdev commented on src/pybind/mgr/dashboard_v2/controllers/cephfs.py:
```diff
@@ -0,0 +1,322 @@
+# -*- coding: utf-8 -*-
+from __future__ import absolute_import
+
+import collections
```
Remove this import, it's not used.

tspmelo replied:
Done.

votdev commented on src/pybind/mgr/dashboard_v2/controllers/cephfs.py:
```diff
@@ -0,0 +1,322 @@
+import cherrypy
+from mgr_module import CommandResult
+
+from ..tools import ApiController, AuthRequired, BaseController, NotificationQueue, ViewCache
```
Remove ``NotificationQueue``, it's not needed

tspmelo replied:
Done.

votdev commented on src/pybind/mgr/dashboard_v2/controllers/cephfs.py:
```diff
@@ -0,0 +1,322 @@
+import collections
+from collections import defaultdict
+import json
+import time
```
Remove this import, it's not used.

tspmelo replied:
Done.

votdev commented on src/pybind/mgr/dashboard_v2/controllers/cephfs.py:
```diff
@@ -0,0 +1,322 @@
+from ..tools import ApiController, AuthRequired, BaseController, NotificationQueue, ViewCache
+
+
+LOG_BUFFER_SIZE = 30
```
Useless?

tspmelo replied:
Done.

votdev commented on src/pybind/mgr/dashboard_v2/frontend/src/app/ceph/cephfs/cephfs.service.ts:
```diff
@@ -0,0 +1,19 @@
+  constructor(private http: HttpClient) {}
+
+  getCephfs(id) {
+    return this.http.get('/api/cephfs/data/' + id + '/');
```
I suggest to use template-strings here.

tspmelo replied:
Done.

votdev commented on src/pybind/mgr/dashboard_v2/frontend/src/app/ceph/cephfs/cephfs.service.ts:
```diff
@@ -0,0 +1,19 @@
+  }
+
+  getClients(id) {
+    return this.http.get('/api/cephfs/clients/' + id + '/');
```
I suggest to use template-strings here.

tspmelo replied:
Done.

votdev commented on src/pybind/mgr/dashboard_v2/frontend/src/app/ceph/cephfs/cephfs.service.ts:
```diff
@@ -0,0 +1,19 @@
+  }
+
+  getMdsCounters(id) {
+    return this.http.get('/api/cephfs/mds_counters/' + id + '/');
```
I suggest to use template-strings here.

tspmelo replied:
Done.

LenzGr commented:
@votdev I agree that we should look into reviewing and improving the workflows and usability of existing pages, but I suggest to postpone "bigger" changes to the UI to after the initial PR has been accepted.

ricardoasmarques commented:
Should't we also have a "title" on the standby table?

![screenshot from 2018-02-19 13-39-11](https://user-images.githubusercontent.com/14297426/36380659-fe643f52-157a-11e8-9c02-402ae7d54061.png)

ricardoasmarques commented:
Maybe you will increase the usability if you put the link on text:

![screenshot from 2018-02-19 13-54-19](https://user-images.githubusercontent.com/14297426/36381119-7cd39774-157c-11e8-9b59-5577a08ea27d.png)

instead of having a link with the label "Detail...":

![screenshot from 2018-02-19 13-54-33](https://user-images.githubusercontent.com/14297426/36381120-7fa79ba8-157c-11e8-8799-e154e962e97d.png)

On the current implementation, user will have to think to answer the question "Detail of what?" and the UI is less clean, IMO.

tspmelo commented:
@ricardoasmarques I reworked the standby table to reduce the screen space it uses, and removed the title for the same reason.
ATM that table will always have just 1 row and the user will easily know what it is showing.

tspmelo commented:
Updated the clients link as recommended by @ricardoasmarques.

ricardoasmarques reviewed (approved):
lgtm

LenzGr reviewed (approved):
LGTM
